File: client/src/presse-papier.ts
// Le presse-papier reçu de la VM, côté navigateur : quoi écrire, quand, et
// quoi dire quand ça ne marche pas.
//
// **PUR — aucun `document`, aucun `navigator`, aucune promesse.** Ce module
// décide ; c'est `presse-papier-dom.ts` qui appelle `writeText` — reçue par
// injection, depuis `main.ts` — et qui lui rapporte le résultat (`confirmer`,
// `echouer`). C'est le patron de `status.ts` et de `resize.ts`, et c'est ce
// qui le rend éprouvable sans DOM.
//
// ⚠️ **Cette phrase disait « c'est `main.ts` qui […] lui rapporte le
// résultat », et c'était vrai quand elle a été écrite.** L'extraction de
// `presse-papier-dom.ts` — jouée AVANT l'addition, dans la même branche — a
// déplacé le câblage : `main.ts` ne construit plus `PressePapierLocal` et ne
// lui rapporte plus rien (`grep -c PressePapierLocal client/src/main.ts`
// rend **0**). Corrigée par la revue transverse du 20 août 2026. C'est le
// mode de défaillance dominant de ce dépôt : une affirmation devenue fausse
// **dans sa propre branche**.
//
// Il n'importe rien de `proto/ts/control.ts` : il prend une `Recu` locale.
// C'est délibéré — le découpler du protocole est ce qui le garde pur, et un
// changement de forme du message ne doit pas traverser jusqu'ici.

/// Ce que l'agent a annoncé.
export interface Recu {
    /// Le texte à écrire, ou `null` quand l'agent a REFUSÉ le contenu parce
    /// qu'il dépassait sa borne. `null` n'est pas « rien » : c'est un refus,
    /// et il se dit.
    texte: string | null;
    /// La taille en octets — celle du texte émis, ou celle du contenu refusé.
    octets: number;
}

/// Ce que le message d'échec doit porter : COMMENT rétablir, pas seulement
/// qu'il manque quelque chose. Même règle que `DETAIL_REFUS` du micro
/// (`micro.ts`), et pour la même raison — un message qui ne dit que le
/// symptôme laisse l'utilisateur sans geste à faire.
export const MESSAGE_ECHEC =
    "copie de la VM non recopiée ici — cliquez dans la fenêtre pour lui rendre le focus, puis recopiez";

/// Nombre d'échecs CONSÉCUTIFS avant de crier.
///
/// **Deux, pas un** : un premier échec est le cas ordinaire d'une fenêtre qui
/// n'a pas le focus au moment où l'agent pousse, et crier là-dessus ferait un
/// bandeau permanent sur un produit qui marche.
export const ECHECS_AVANT_MESSAGE = 2;

/// Taille maximale, en **octets d'UTF-8**, d'un texte que la page accepte
/// d'émettre vers l'agent (sous-bloc P2).
///
/// 🔴 **C'EST UNE COPIE, ET RIEN DANS LE LANGAGE NE LA CONFRONTE À SA
/// SOURCE.** La valeur qui fait foi est `agent::presse_papier::PRESSE_PAPIER_MAX`
/// (`agent/src/presse_papier.rs`), et `client/` ne peut pas importer de Rust.
/// Le dépôt a déjà payé cette classe — deux constantes écrites dans deux
/// langages sans `import` possible divergent en SILENCE (sous-bloc P2 de la
/// plateforme). Le remède employé est le même qu'alors : **un test qui relit
/// le fichier Rust et refuse la divergence**, dans `presse-papier.test.ts`.
///
/// **Pourquoi la borne est ici et pas seulement chez l'agent** : sans elle,
/// l'agent la ferait bien respecter, mais le canal de contrôle aurait DÉJÀ
/// porté la charge, et le bandeau ne paraîtrait jamais — l'agent refuse en
/// journalisant, sans rien renvoyer. C'est ici, et ici seulement, que
/// l'utilisateur peut être averti.
export const PRESSE_PAPIER_MAX = 64 * 1024;

/// Le message de refus, qui NOMME la taille — « trop grand » seul ne dit pas
/// à l'utilisateur ce qu'il doit réduire.
export function messageDeRefus(octets: number): string {
    const kio = Math.round(octets / 1024);
    return `copie trop volumineuse (${kio} Kio) — elle n'a pas été recopiée ici, réduisez la sélection`;
}

export class PressePapierLocal {
    /// Le dernier texte reçu et pas encore écrit. **Un seul**, jamais une
    /// file : une écriture obsolète est impossible parce qu'on ne garde que
    /// le dernier.
    private enAttente: string | undefined;
    /// Le dernier texte réellement écrit — on ne le réécrit pas.
    private ecrit: string | undefined;
    /// Échecs consécutifs d'écriture.
    private echecs = 0;
    /// Le refus à dire, **consommable** : sinon le bandeau se réafficherait à
    /// chaque tour.
    private refus: string | undefined;
    /// Le dernier texte REÇU de l'agent et pas encore réémis — le **garde n°3
    /// de D5**, et il se **consomme**.
    ///
    /// Il est posé par `recevoir`, jamais par `confirmer` : un texte reçu sans
    /// focus reste en attente d'écriture, et l'utilisateur peut coller
    /// entre-temps. Les deux cas sont indiscernables de l'extérieur, et le
    /// choix est de se taire — un aller-retour évité de trop coûte un collage
    /// répété que l'utilisateur peut refaire, là où un aller-retour de trop
    /// est du trafic que rien ne borne (legs n°4 de P1).
    private recuNonReemis: string | undefined;

    /// Un message est arrivé de l'agent. **Toujours mémorisé**, même sans
    /// focus : c'est le dépôt différé.
    recevoir(recu: Recu): void {
        if (recu.texte === null) {
            // Un refus n'écrase PAS le dernier texte mémorisé : sinon il
            // effacerait un contenu valide encore non écrit.
            this.refus = messageDeRefus(recu.octets);
            return;
        }
        this.enAttente = recu.texte;
        // Arme le garde n°3 : ce texte-là ne repartira pas vers l'agent.
        this.recuNonReemis = recu.texte;
    }

    /// Ce qu'il faut écrire MAINTENANT, ou `undefined`.
    ///
    /// Sans focus on ne rend rien : `navigator.clipboard.writeText` échoue
    /// sur un document qui n'a pas le focus, et l'échec coûterait un compteur
    /// pour rien. Le texte reste en attente et sortira au retour du focus.
    ///
    /// ⚠️ **CETTE PHRASE AFFIRMAIT COMME UN FAIT CE QUE LA SPEC DÉCLARE
    /// SUPPOSÉ DEPUIS LE 28 JUILLET 2026** (§3.3). Le sous-bloc P3 l'a
    /// mesurée, et le verdict est plus fin que « vrai » ou « faux » :
    ///
    /// - la cellule qui TRANCHE — pas de focus, MAIS sous activation
    ///   utilisateur — est **INATTEIGNABLE** à ce montage : le geste de
    ///   confiance REND le focus à la fenêtre qui le reçoit, et
    ///   `Page.bringToFront` ne le lui reprend plus. Le §3.3 reste donc
    ///   **supposé au sens strict** ;
    /// - **mais il est CORROBORÉ par une pièce** : sans focus et sans geste,
    ///   `writeText` refuse en NOMMANT le focus —
    ///   `NotAllowedError: … Document is not focused.` — là où le refus
    ///   d'activation dit `… Write permission denied.` Les deux portent le
    ///   MÊME NOM et des MESSAGES DIFFÉRENTS : **un chemin de refus propre au
    ///   focus existe, et il se nomme lui-même**. Ce qui reste non mesuré est
    ///   s'il survit à une activation.
    ///
    /// 🔵 **ET À N FENÊTRES, CE TEST FAIT AUTRE CHOSE QUE SE PROTÉGER D'UN
    /// REFUS — il ÉLIT l'unique écrivain local.** Le capteur pousse le contenu
    /// à TOUTES les fenêtres (D3), chacune a son propre `PressePapierLocal`,
    /// et si toutes écrivaient, N appels concurrents à `writeText` partiraient
    /// pour une seule copie, le dernier gagnant arbitrairement. Cette
    /// justification-là vaut **indépendamment** du §3.3, et c'est pourquoi la
    /// règle reste même si le §3.3 devait être réfuté un jour. Le retrait de
    /// la règle serait alors une **décision du propriétaire du dépôt**, avec
    /// son coût nommé — un régime que rien ne mesure —, jamais une conséquence
    /// mécanique d'un verdict de sonde.
    ///
    /// ⚠️ **Sondes versées** : `journaux-presse-papier-p3/p3-writetext-{1,2}.json`
    /// (le 2×2) et `p3-focus-{1,2}.json` (la mesurabilité du focus à N
    /// fenêtres), deux exécutions chacune, relevés identiques.
    aEcrire(focalise: boolean): string | undefined {
        if (!focalise) return undefined;
        if (this.enAttente === undefined) return undefined;
        if (this.enAttente === this.ecrit) return undefined;
        return this.enAttente;
    }

    /// L'écriture a réussi.
    confirmer(texte: string): void {
        this.ecrit = texte;
        // Un succès remet le compteur à zéro : sans cela un échec au démarrage
        // et un échec une heure plus tard crieraient ensemble.
        this.echecs = 0;
    }

    /// L'écriture a échoué. Rend le message à afficher au DEUXIÈME échec
    /// consécutif, `undefined` avant.
    echouer(): string | undefined {
        this.echecs += 1;
        return this.echecs >= ECHECS_AVANT_MESSAGE ? MESSAGE_ECHEC : undefined;
    }

    /// Rend le texte à émettre vers l'agent, ou `undefined` si c'est l'écho
    /// d'un contenu qu'on vient de recevoir de lui — **le garde n°3 de D5**.
    ///
    /// **Il ne vaut que pour le PREMIER renvoi**, et c'est délibéré : un
    /// utilisateur qui colle deux fois le même texte le veut deux fois. Le
    /// doublon n'en coûte rien côté VM — le garde n°2 de l'agent l'absorbe,
    /// le presse-papier Windows portant déjà ce contenu.
    ///
    /// ⚠️ **Un refus n'arme pas ce garde** : rien n'a été écrit localement,
    /// donc rien ne peut en être l'écho.
    aEmettre(texte: string): string | undefined {
        const recu = this.recuNonReemis;
        this.recuNonReemis = undefined;
        return recu === texte ? undefined : texte;
    }

    /// Le refus à dire, ou `undefined`. **Se consomme.**
    refusADire(): string | undefined {
        const refus = this.refus;
        this.refus = undefined;
        return refus;
    }
}

File: agent/src/presse_papier.rs
//! Le presse-papier de la VM, sens **VM → navigateur** : détecter qu'il a
//! changé, en lire le texte, et décider ce qu'on annonce.
//!
//! **Pur, sans aucun `cfg`** — comme `capteur/plein_ecran.rs`,
//! `capteur/audio.rs` et `capteur/repartiteur.rs` avant lui. Toute la
//! décision vit ici et s'éprouve sur l'hôte Linux ; les deux appels Win32
//! vivent dans `presse_papier/win32.rs`, gaté, et **ne décident rien**.
//!
//! Le module est à la **racine nue** (`mod presse_papier;` dans `main.rs`) et
//! non sous `capteur/`, alors que le propriétaire est aujourd'hui le capteur
//! et lui seul. La raison n'est pas celle que la spec avance — la « Convention
//! de module enfant » de `CLAUDE.md` déclare elle-même sa portée et ne couvre
//! pas ce cas, ce module n'étant extrait de rien. C'est que la décision D1 pose
//! que le propriétaire est « le capteur quand il existe, l'enfant sinon » : un
//! module rangé sous `capteur/` porterait un nom faux le jour où le
//! propriétaire mono-fenêtre arrivera.
//!
//! ❌ **Ce module disait « il n'écrit JAMAIS le presse-papier Windows ; le sens
//! navigateur → VM est le sous-bloc P2, et c'est lui qui portera les gardes
//! anti-écho de D5 ; le seul garde livré ici est le n°2 ». LES TROIS CLAUSES
//! SONT PÉRIMÉES : ce sous-bloc a eu lieu.** Relevé par la revue transverse du
//! 21 août 2026. La règle qu'il énonçait tient toujours, mais autrement :
//!
//! - **il n'écrit toujours pas lui-même** — l'appel Win32 vit dans
//!   `presse_papier/win32.rs`, gaté, et `ecrire_la_plateforme` n'est que
//!   l'aiguillage de plateforme, jumeau de `Sondeur::lire_la_plateforme` ;
//! - **les gardes n°1 et n°2 sont ici**, tous deux posés par
//!   `Sondeur::apres_notre_ecriture` sur NOTRE PROPRE écriture. Le n°3 vit
//!   côté page (`client/src/presse-papier.ts`), le seul endroit d'où un écho
//!   pourrait repartir ;
//! - **le n°2 absorbe toujours le faux positif du compteur** — celui-ci **bouge
//!   sur une réécriture identique**, mesuré (sonde P0, `q2="bouge"`, deux
//!   exécutions du 20 août 2026) — mais il ferme désormais AUSSI l'aller-retour
//!   d'un collage, ce que P1 ne pouvait pas produire.
//!
//! ⚠️ **Et « il ne ferme aucune boucle (il n'y en a pas) » reste VRAI**, contre
//! toute attente : dans l'architecture livrée, aucune oscillation
//! auto-entretenue n'est possible, chaque tour exigeant un geste humain — le
//! client n'émet vers l'agent que sur un `paste`. Ce que les gardes suppriment
//! est **un aller-retour par collage**, pas une divergence. Voir `gardes_armes`,
//! qui porte la démonstration et la conséquence sur la rouge du critère ④.

use std::time::Duration;

/// Taille maximale, en octets d'UTF-8 **après normalisation**, d'un contenu
/// que l'on accepte de pousser au navigateur.
///
/// ⚠️ **NON CALIBRÉE.** Elle rejoint `BPP_MIN`, `FACTEUR_FOCUS`,
/// `PART_DORMANTE_BPS`, `HYSTERESIS` et `TAILLE_MAX_SORTIE` : aucun jugement
/// d'usage n'a été porté sur sa valeur. 64 KiB tient un document texte
/// ordinaire et refuse un presse-papier chargé d'un fichier entier.
///
/// **Au-delà, on REFUSE — on ne tronque pas.** Un collage silencieusement
/// amputé est le pire résultat possible, et il est pire que pas de collage du
/// tout : l'utilisateur ne peut pas voir qu'il lui manque la fin.
pub const PRESSE_PAPIER_MAX: usize = 64 * 1024;

/// Période minimale entre deux lectures du compteur de séquence.
///
/// ⚠️ **Ce n'est PAS `PERIODE_REARBITRAGE`**, qui cadence le tour de roue du
/// registre de sommeil. La valeur est du même ordre, délibérément, mais la
/// constante est propre à ce module : les faire suivre l'une l'autre
/// coupleraient deux mécanismes que rien ne lie — c'est exactement l'argument
/// que `plein_ecran::PERIODE_STYLE` porte déjà pour la relecture du style.
///
/// `Sondeur::tour` porte donc son propre minuteur et rend `None` sans rien
/// lire tant qu'il n'est pas échu, **même si le tour de roue l'appelle plus
/// souvent**.
pub const PERIODE_PRESSE_PAPIER: Duration = Duration::from_millis(250);

/// Ce que le sondeur a décidé d'annoncer.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Annonce {
    /// Le texte à pousser, **déjà normalisé et sous la borne**.
    Texte(String),
    /// Un contenu de `octets` octets d'UTF-8, **après normalisation**, a été
    /// REFUSÉ — jamais tronqué. Le compte sert au bandeau côté client, qui
    /// doit pouvoir dire *combien* plutôt que « trop grand ».
    Refus { octets: u32 },
}

/// `PRESSE_PAPIER=0` désarme le mécanisme entier.
///
/// **`=0` DÉSACTIVE, une simple présence n'active pas**, exactement comme
/// `PLEIN_ECRAN`, `AUDIO`, `SUPERVISEUR` et `CAPTEUR` : tester `is_ok()`
/// armerait le mécanisme en écrivant `PRESSE_PAPIER=0` pour le couper.
///
/// `OnceLock` et non une lecture par appel : le sondage court à 4 Hz, et
/// l'environnement ne change pas en cours de processus.
pub fn actif() -> bool {
    static ACTIF: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
    *ACTIF.get_or_init(|| {
        let actif = std::env::var("PRESSE_PAPIER").as_deref() != Ok("0");
        if !actif {
            tracing::warn!(
                "presse-papier DESARME (PRESSE_PAPIER=0) : le contenu copie dans la VM \
                 n'est plus pousse au navigateur"
            );
        }
        actif
    })
}

/// `PRESSE_PAPIER_GARDE=0` désarme les gardes anti-écho de `apres_notre_ecriture`.
///
/// ⚠️ **VARIABLE DE BANC, JAMAIS UNE CONFIGURATION LIVRÉE** — même statut que
/// `PART_SONDAGE`. Elle existe pour un seul usage : rendre ATTEIGNABLE la rouge
/// du critère ④ de P2, qui compte les messages `clipboard` revenant vers la
/// fenêtre après un collage.
///
/// **`=0` DÉSARME ; une simple présence n'arme pas.** Les gardes sont armés par
/// défaut, et tester `is_ok()` les désarmerait en écrivant
/// `PRESSE_PAPIER_GARDE=0` pour... les désarmer. Convention de `PLEIN_ECRAN`,
/// `AUDIO`, `SUPERVISEUR`, `CAPTEUR`, `PART_SONDAGE` et `PRESSE_PAPIER`.
///
/// 🔴 **ELLE DÉSARME LES DEUX GARDES, PAS LE SEUL N°1, ET C'EST LE POINT.**
/// La spécification prescrivait de désarmer le n°1 et d'attendre un compte qui
/// « croît sans borne » ; **il reste à un, et la spec avait prévu ce cas**.
/// Sans armement du n°2, le `Sondeur` relit notre texte, l'annonce **une**
/// fois, puis pose lui-même `dernier_emis` et `reference` — au tour suivant
/// `observer` sort sur sa première ligne. Et rien ne relance : le client
/// n'émet vers l'agent que sur un `paste`, donc sur un GESTE HUMAIN, jamais à
/// la réception d'un `clipboard`. Désarmer le seul n°1 rendrait donc **zéro
/// message aussi**, et la rouge serait vacueuse une seconde fois.
///
/// 🔵 **Conséquence de conception, et elle contredit une phrase de D5** : dans
/// l'architecture livrée, **aucune oscillation auto-entretenue n'est
/// possible**, chaque tour exigeant un geste humain. Ce que les gardes
/// suppriment est **un aller-retour par collage**, pas une divergence.
/// ⚠️ Déduit du code, pas d'une mesure : `client/src/presse-papier-dom.ts`
/// n'écrit que localement à la réception et n'émet rien. La condition qui
/// rendrait la boucle réelle est nommée — un client qui réémettrait ce qu'il
/// reçoit —, et c'est précisément ce que le garde n°3 empêche côté page.
pub(super) fn gardes_armes() -> bool {
    static ARMES: std::sync::OnceLock<bool> = std::sync::OnceLock::new();
    *ARMES.get_or_init(|| {
        let armes = std::env::var("PRESSE_PAPIER_GARDE").as_deref() != Ok("0");
        if !armes {
            tracing::warn!(
                "garde anti-echo du presse-papier DESARME (PRESSE_PAPIER_GARDE=0) : \
                 bras de banc, jamais une configuration livree"
            );
        }
        armes
    })
}

/// Ramène toutes les fins de ligne à `\n`.
///
/// Windows écrit `\r\n` ; d'anciennes applications écrivent un `\r` **seul**.
/// Les deux doivent devenir `\n`, sans quoi l'aller-retour de P2 doublerait
/// les lignes à chaque tour. La fonction est **idempotente** : la rejouer sur
/// son propre résultat ne change rien.
pub fn normaliser(texte: &str) -> String {
    let mut sortie = String::with_capacity(texte.len());
    let mut precedent_cr = false;
    for c in texte.chars() {
        match c {
            '\r' => {
                sortie.push('\n');
                precedent_cr = true;
            }
            '\n' => {
                // Le `\n` d'un `\r\n` a déjà été rendu par le `\r`.
                if !precedent_cr {
                    sortie.push('\n');
                }
                precedent_cr = false;
            }
            autre => {
                sortie.push(autre);
                precedent_cr = false;
            }
        }
    }
    sortie
}

/// `\n` → `\r\n`, la réciproque de `normaliser`. **Windows attend `\r\n`.**
///
/// Elle n'est PAS un `replace("\n", "\r\n")` : le texte qui arrive du
/// navigateur peut porter DÉJÀ des `\r\n` — un copier depuis un éditeur
/// Windows local en porte —, et le remplacement naïf rendrait alors `\r\r\n`,
/// donc une ligne vide de plus à chaque collage. La fonction est **idempotente**
/// exactement comme `normaliser` l'est dans l'autre sens, et l'aller-retour
/// `normaliser(denormaliser(x)) == x` est ce qu'un test doit voir rouge en
/// premier (spec §7.1).
pub fn denormaliser(texte: &str) -> String {
    let mut sortie = String::with_capacity(texte.len() + texte.len() / 16);
    let mut precedent_cr = false;
    for c in texte.chars() {
        match c {
            '\r' => {
                sortie.push_str("\r\n");
                precedent_cr = true;
            }
            '\n' => {
                // Le `\n` d'un `\r\n` a déjà été rendu par le `\r`.
                if !precedent_cr {
                    sortie.push_str("\r\n");
                }
                precedent_cr = false;
            }
            autre => {
                sortie.push(autre);
                precedent_cr = false;
            }
        }
    }
    sortie
}

/// Borne le texte ENTRANT, en octets d'UTF-8. **On REFUSE, on ne tronque pas.**
///
/// ⚠️ **Ce n'est pas la même borne que celle du sens sortant, et l'asymétrie
/// est voulue.** Côté sortant, `PRESSE_PAPIER_MAX` protège le **canal de
/// contrôle** (D4) : le texte n'y est pas encore passé. Côté entrant, le texte
/// a **déjà** traversé ce canal quand l'agent le voit — la borne y protège le
/// tube capteur↔enfant et la mémoire, pas le canal. C'est le client qui doit
/// appliquer la sienne AVANT d'émettre ; celle-ci est la ceinture.
///
/// **Le refus entrant se journalise et ne remonte aucun bandeau** : le client a
/// déjà refusé et dit pourquoi, et un second bandeau pour le même geste serait
/// du bruit.
///
/// La borne porte sur `len()`, c'est-à-dire des **octets**, jamais sur
/// `chars().count()` : c'est l'unité du canal, et un texte d'emojis dont le
/// compte de caractères tient déborde de quatre fois en octets.
pub fn borner_entrant(texte: &str) -> Option<String> {
    (texte.len() <= PRESSE_PAPIER_MAX).then(|| texte.to_owned())
}


/// Écrit le presse-papier de la VM, et rend le numéro de séquence relu APRÈS
/// la fermeture — celui qu'il faut passer à `Sondeur::apres_notre_ecriture`.
///
/// **Jumelle exacte de `Sondeur::lire_la_plateforme`**, et posée au même
/// endroit pour la même raison : l'appelant (`capteur/sommeil/presse_papier.rs`)
/// n'est pas gaté et doit compiler sur l'hôte Linux.
///
/// ⚠️ **Le texte doit arriver DÉJÀ dénormalisé** (`\r\n`) : cette fonction ne
/// décide rien, elle transmet.
#[cfg(windows)]
pub fn ecrire_la_plateforme(texte: &str) -> anyhow::Result<u32> {
    win32::ecrire_texte(texte)
}

/// Repli non-Windows. **Un `Err`, jamais un `Ok`** : rendre `Ok(0)` ferait
/// croire à un succès, et l'appelant injecterait `Ctrl+V` sur un
/// presse-papier inchangé — c'est-à-dire collerait le contenu PRÉCÉDENT, le
/// mode de défaillance silencieux que D6 existe entièrement pour éviter.
#[cfg(not(windows))]
pub fn ecrire_la_plateforme(_texte: &str) -> anyhow::Result<u32> {
    anyhow::bail!("le presse-papier de la VM n'existe pas hors de Windows")
}

/// `Sondeur`, extrait VERBATIM au sous-bloc P3 (tâche 3), AVANT l'addition qui
/// l'a rendu nécessaire. Un `mod` ORDINAIRE, et non un `#[path]` : la
/// « Convention de module enfant » de `CLAUDE.md` ne vise que les modules
/// extraits d'un parent `#[cfg(windows)]`, et ce fichier n'est pas gaté.
mod sondeur;
pub use sondeur::Sondeur;

#[cfg(windows)]
mod win32;

// Les tests de ce module vivent à part depuis le sous-bloc P2 du chantier
// presse-papier : le fichier était à 428 lignes pour un plafond de 500, et P2 y
// ajoute le garde n°1 de D5, la réciproque de `normaliser` et leurs tests.
// L'extraction précède l'addition, comme la règle du dépôt l'exige.
//
// ⚠️ Cet emploi de `#[path]` est HORS de la portée de la « Convention de module
// enfant » de `CLAUDE.md` : c'est le même mécanisme Rust employé pour une autre
// raison — la règle des 500 lignes —, exactement comme `superviseur/table.rs`.
// Ce module ne se hisse PAS à la racine du crate.
#[cfg(test)]
#[path = "presse_papier/tests.rs"]
mod tests;
